human-controls/Human-Controls.ino

```diff
@@ -1,25 +1,71 @@
+#include "Arduino.h"
 #include "FTDebouncer.h"
-#include "RotaryKnobController.h"
+#include "src/RotaryKnobController.h"
+#include "src/CharacterDisplay.h"
+#include "src/MenuController.h"
+#include "src/DisplayController.h"
+#include "src/DisplayPage.h"
+#include "src/Printer.h"
 
+//Pins
 #define ENCODER_PIN_CLK 3 //CLK gets degrees for rotary knob
 #define ENCODER_PIN_DT 4  //DT gets direction for rotary knob
 #define ENCODER_PIN_SW 5  //Gets the button for rotary knob
 
+//Other constraints
+#define DISPLAY_ADDRESS 0X27 //I2c address of the lcd display
+#define DISPLAY_LENGTH 16    //Length of the lcd display
+#define DISPLAY_WIDTH 2      //width of the lcd display
+#define MENU_SIZE 4          //Number of menu pages
 #define USB_BAUDRATE 115200
 
+//Lengths of printers inside each DisplayPage
+#define DASH_PRINTER_LEN 1
+#define ELEVATOR_PRINTER_LEN 4
+#define SHOT_PRINTER_LEN 4
+#define VALVE_PRINTER_LEN 4
+
+//Min - Max
+#define ANGLE_MIN 0      //Minimum elevator angle
+#define ANGLE_MAX 10     //Maximum elevator angle
+#define PRESSURE_MIN 60  //Minimum shot pressure
+#define PRESSURE_MAX 100 //Maximum shot pressure
+#define DURATION_MIN 0   //Minimum valve duration
+#define DURATION_MAX 10  //Maximum valve duration
+
 FTDebouncer pinDebouncer(30);
-RotaryKnobController rotaryKnob(ENCODER_PIN_CLK, ENCODER_PIN_DT);
+//RotaryKnobController rotaryKnob(ENCODER_PIN_CLK, ENCODER_PIN_DT);
+
+Printer dashPrinters[DASH_PRINTER_LEN]{{2, 0, "Dash", false}};
+Printer elevatorPrinters[ELEVATOR_PRINTER_LEN]{{2, 0, "Elevator Angle", false}, {6, 1, "<[", true}, {8, 1, "!modifyValue!", false}, {10, 1, "]>", true}};
+Printer shotPrinters[SHOT_PRINTER_LEN]{{2, 0, "Shot Pressure", false}, {6, 1, "<[", true}, {8, 1, "!modifyValue!", false}, {10, 1, "]>", true}};
+Printer valvePrinters[VALVE_PRINTER_LEN]{{2, 0, "Valve Duration", false}, {6, 1, "<[", true}, {8, 1, "!modifyValue!", false}, {10, 1, "]>", true}};
```

Comment:
Have each page create its own printers in its constructor, then have the MenuController instantiate the pages in its constructor.

Reply:
Done